### ISSUE TITLE
auto_arima and multiprocessing

### ISSUE THREAD

**Issue by Juni0rMafia** [tgsmith61591/pmdarima#120]
Hello, 

Thank you very much for your pmdarima auto_arima very useful function. I have been using it intensively lately and ended with models taking hours to compute, due to multiple auto_arima calls. Therefore, I added multiprocessing to the show. 

I simplified my code, you can see it below.
With NO multiprocessing involved, each auto_arima's total fit time takes around **8 seconds**.
With multiprocessing involved, the auto_arima functions are called 4 by 4 (I have 4 cores), in theory it works.
But the results are disappointing, each auto_arima's total fit time is about **20 seconds** with the data plugged.  Therefore I don't gain much time using multiprocessing.

Is the auto_arima somehow incompatible with multiprocessing ?

Thank you very much
```python
import os
import time
import multiprocessing as mp
import pandas as pd
from pmdarima.arima import auto_arima
    
def myFunction(data):
    data_Monthly_A = data.resample('M').mean()
    data_Monthly_A.dropna(inplace = True)
    data_Monthly_A.columns=['data_Monthly_A']
    data_Monthly_A = data_Monthly_A.iloc[0:100,:]
    
    model = auto_arima(data_Monthly_A, start_p=1, start_q=1,
                                       max_p=3, max_q=3, m=12,
                                       start_P=0, seasonal=True,
                                       d=1, D=1, trace=True,
                                       error_action='ignore',  
                                       suppress_warnings=True, 
                                       stepwise=True)
    return model

data = getDatafromFile()
data_ = [None]*10

for i in range(0,10):
    data_[i]=data 

startTime = time.time()
pool = mp.Pool(mp.cpu_count())
asyncResult = pool.map_async(myFunction, data_)	
resultList = asyncResult.get()
print(time.time() - startTime)
```

**Comment by tgsmith61591** [comment]
Hey there, thanks for the extremely well-formatted issue. I'm going to take a look into this over the weekend, and hope to be able to shed some light on it soon.

**Comment by vibhashahani** [comment]
I am facing the same issue.
Is auto_arima compatible with multiprocessing?

```
smape = []
model= []

m  = auto_arima(self.df_train, start_p=0, start_q=0,
                            max_p=5, max_q=5, m=self.season,
                            start_P=0, start_Q=0,
                            seasonal=True,
                            max_P=3, max_Q=3,
                            d=1, D=1,
                            max_order = 10,
                            trace=False,
                            suppress_warnings = True,
                            error_action = 'ignore',
                            stepwise = True)

m.fit(self.df_train)

with multiprocessing.Pool(processes=4) as pool:
            results = pool.starmap(self.arima_model, par)
        pool.close()
        pool.join()

        for x in results:
            smape.append(x["smape"])
            model.append(x["model"])

```

**Comment by MattConflitti** [comment]
Any recent updates on or insights into this issue?

**Comment by tgsmith61591** [comment]
I haven't had much time to dig into this, sorry. I threw together a script this morning and I'm unable to replicate. Here's what I've got:

```python
import os
import time
import ctypes
from multiprocessing import Pool, cpu_count, Array
import pmdarima as pm
import numpy as np

# I don't have your dataset, so i'll just use an internal dataset
data = pm.datasets.load_wineind()


def timed(func):
    def _wrapper(*args, **kwargs):
        start = time.time()
        res = func(*args, **kwargs)
        print(f"Completed {func.__name__} in {time.time() - start:.3f} sec")
        return res
    return _wrapper


def _fit_func(y):
    return pm.auto_arima(
        y, seasonal=True, m=12, trace=False, error_action='ignore',
        suppress_warnings=True, stepwise=True,
        sarimax_kwargs={'simple_differencing': True})  # new in 1.3.0


@timed
def fit_serial(n):
    fits = []
    for i in range(n):
        fits.append(_fit_func(data))
    return fits


@timed
def fit_parallel_nv(n):
    pool = Pool(cpu_count())
    res = pool.map_async(_fit_func, (data for i in range(n)))
    return res.get()


# for shared memory
shared_array_base = Array(ctypes.c_double, data)
shared_array = np.ctypeslib.as_array(shared_array_base.get_obj())

def _fit_func_shared(i, y=shared_array):
    print(f"fitting #{i}")
    return pm.auto_arima(
        y, seasonal=True, m=12, trace=False, error_action='ignore',
        suppress_warnings=True, stepwise=True,
        sarimax_kwargs={'simple_differencing': True})


@timed
def fit_parallel_shared(n):
    pool = Pool(cpu_count())
    res = pool.map_async(_fit_func_shared, (i for i in range(n)))
    return res.get()


if __name__ == "__main__":
    serial = fit_serial(5)
    # Completed fit_serial in 42.271 sec

    parallel_naive = fit_parallel_nv(5)
    # Completed fit_parallel_nv in 24.737 sec

    parallel_shared = fit_parallel_shared(5)
    # Completed fit_parallel_shared in 25.607 sec
```

As you can see, a serial fit takes 42 seconds, an embarrassingly parallelized approach (with the data being pickled for each process) takes 24, and a shared memory approach takes 25. I'd call both parallel methods roughly comparable, though if the data were any larger, I'd imagine the shared memory approach would begin to shine.

### tl;dr

This should be completely compatible with multiprocessing (so long as you're using `stepwise=True`). I'm not sure why you're seeing the slow down.

### Speculation

In multiprocessing, every function and arg will be pickled in order for a process to pick them up and execute them. All I can really do is speculate here... perhaps the data you're sending into the function is so large that you're spending lots of time in pickling overhead?

Moreover, I struggle to understand the objective here. Maybe it's just a toy example, but you're basically fitting the same exact model over the same dataset 10 times (in parallel). Am I wrong?

**Comment by MattConflitti** [comment]
The pickling overheard may be spot on. I am not sure of OP's use case, but my goal is to be able to fit multiple models for thousands of vastly different datasets. I then can pick and choose which models perform the best for an upcoming forecast across those datasets. I need to parallelize to get this done in a reasonable amount of time.

Also, does the `simple_differencing` argument help with speed at all?

**Comment by tgsmith61591** [comment]
`simple_differencing` will definitely help with seasonal speed:

> Whether or not to use partially conditional maximum likelihood estimation. If True, differencing is performed prior to estimation, which discards the first 𝑠𝐷+𝑑 initial rows but results in a smaller state-space formulation. If False, the full SARIMAX model is put in state-space form so that all datapoints can be used in estimation. Default is False.

Since I can't reproduce this, I'm going to close unless a more reproducible example is provided

**Comment by vinson2233** [comment]
The issue is auto_arima is using all cores even with njobs = 1. Thats what multiprocessing with auto arima is slow. This thread have the answer https://github.com/alkaline-ml/pmdarima/issues/301 for this issue.
